From 9179cb58caa3648095c6d108a1cff7651b5befe5 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 30 Oct 2025 15:02:43 +0800
Subject: [PATCH 089/105] mtk: mt76: mt7996: introduce chanctx state

Introduce chanctx state so that we can set some MCU commands according
to the current state.

---
 channel.c     | 28 +++++++++++++++++++++++++++-
 mt76.h        | 11 +++++++++++
 mt7996/main.c | 36 +++++++++++++++++++++++++++++++++++-
 3 files changed, 73 insertions(+), 2 deletions(-)

diff --git a/channel.c b/channel.c
index 7b4950bfd..73fa24378 100644
--- a/channel.c
+++ b/channel.c
@@ -48,6 +48,8 @@ int mt76_add_chanctx(struct ieee80211_hw *hw,
 		mt76_abort_scan(dev);
 
 	mutex_lock(&dev->mutex);
+
+	ctx->state = MT76_CHANCTX_STATE_ADD;
 	if (!phy->chanctx)
 		ret = mt76_phy_update_channel(phy, conf);
 	else
@@ -77,8 +79,10 @@ void mt76_remove_chanctx(struct ieee80211_hw *hw,
 		mt76_abort_scan(dev);
 
 	mutex_lock(&dev->mutex);
-	if (phy->chanctx == ctx)
+	if (phy->chanctx == ctx) {
 		phy->chanctx = NULL;
+		phy->radar_enabled = false;
+	}
 	mutex_unlock(&dev->mutex);
 }
 EXPORT_SYMBOL_GPL(mt76_remove_chanctx);
@@ -101,6 +105,7 @@ void mt76_change_chanctx(struct ieee80211_hw *hw,
 	cancel_delayed_work_sync(&phy->mac_work);
 
 	mutex_lock(&dev->mutex);
+	ctx->state = MT76_CHANCTX_STATE_CHANGE;
 	mt76_phy_update_channel(phy, conf);
 	mutex_unlock(&dev->mutex);
 }
@@ -260,6 +265,27 @@ int mt76_switch_vif_chanctx(struct ieee80211_hw *hw,
 
 skip_link_replace:
 	for (i = 0; i < n_vifs; i++) {
+		old_ctx = (struct mt76_chanctx *)vifs[i].old_ctx->drv_priv;
+		new_ctx = (struct mt76_chanctx *)vifs[i].new_ctx->drv_priv;
+
+		new_ctx->state = MT76_CHANCTX_STATE_SWITCH;
+
+		if (vifs[i].vif->type == NL80211_IFTYPE_AP)
+			new_ctx->has_ap = true;
+		else if (vifs[i].vif->type == NL80211_IFTYPE_STATION)
+			new_ctx->has_sta = true;
+
+		mt76_dbg(dev, MT76_DBG_CHAN,
+			 "%s: chan=%d->%d, width=%d->%d, punct_bitmap=0x%04x->0x%04x, link=%u\n",
+			 __func__,
+			 vifs[i].old_ctx->def.chan->hw_value,
+			 vifs[i].new_ctx->def.chan->hw_value,
+			 vifs[i].old_ctx->def.width,
+			 vifs[i].new_ctx->def.width,
+			 vifs[i].old_ctx->def.punctured,
+			 vifs[i].new_ctx->def.punctured,
+			 vifs[i].link_conf->link_id);
+
 		mlink = mt76_vif_conf_link(dev, vifs[i].vif, vifs[i].link_conf);
 		if (!mlink)
 			continue;
diff --git a/mt76.h b/mt76.h
index 3a7652442..de65719eb 100644
--- a/mt76.h
+++ b/mt76.h
@@ -1464,8 +1464,19 @@ struct mt76_ethtool_worker_info {
 	int sta_count;
 };
 
+enum mt76_chanctx_state {
+	MT76_CHANCTX_STATE_UNSPEC,
+	MT76_CHANCTX_STATE_ADD,
+	MT76_CHANCTX_STATE_CHANGE,
+	MT76_CHANCTX_STATE_SWITCH,
+};
+
 struct mt76_chanctx {
 	struct mt76_phy *phy;
+
+	enum mt76_chanctx_state state;
+	bool has_ap:1;
+	bool has_sta:1;
 };
 
 #define CCK_RATE(_idx, _rate) {					\
diff --git a/mt7996/main.c b/mt7996/main.c
index 292708417..5a6acb868 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -692,7 +692,41 @@ static void mt7996_remove_interface(struct ieee80211_hw *hw,
 int mt7996_set_channel(struct mt76_phy *mphy)
 {
 	struct mt7996_phy *phy = mphy->priv;
-	int ret;
+	int ret = 0;
+
+	if (mphy->chanctx && mphy->chanctx->state == MT76_CHANCTX_STATE_ADD) {
+		if (!mt76_testmode_enabled(mphy) && !mt76_testmode_bf_enabled(mphy)) {
+			ret = mt7996_mcu_edcca_enable(phy, true);
+			if (ret)
+				goto out;
+		}
+
+		ret = mt7996_mcu_set_pp_en(phy, PP_USR_MODE,
+					   mphy->chandef.punctured);
+		if (ret)
+			goto out;
+	} else if (mphy->chanctx && mphy->chanctx->state == MT76_CHANCTX_STATE_SWITCH) {
+		u8 delta;
+		int current_txpower;
+
+		if (mphy->chanctx->has_ap && phy->pp_mode == PP_USR_MODE) {
+			ret = mt7996_mcu_set_pp_en(phy, PP_USR_MODE,
+						   mphy->main_chandef.punctured);
+		} else if (mphy->chanctx->has_sta) {
+			u8 omac_idx = get_omac_idx(NL80211_IFTYPE_STATION,
+				      phy->omac_mask);
+			ret = mt7996_mcu_set_pp_sta_dscb(phy, &mphy->main_chandef,
+							 omac_idx);
+		}
+		if (ret)
+			goto out;
+
+		delta = mt76_tx_power_path_delta(hweight16(mphy->chainmask));
+		current_txpower = DIV_ROUND_UP(mphy->txpower_cur + delta, 2);
+		ret = mt7996_mcu_set_txpower_sku(phy, current_txpower);
+		if (ret)
+			goto out;
+	}
 
 	if (phy->dev->cal.data) {
 		ret = mt7996_mcu_apply_tx_dpd(phy);
-- 
2.45.2

